qiskit/passmanager/base_tasks.py

```diff
@@ -0,0 +1,227 @@
+
+# Type alias
+PassManagerIR = Any
+
+
+class Task(ABC):
+    """An interface of the pass manager task.
+
+    The task takes a Qiskit IR, and outputs new Qiskit IR after some operation on it.
+    A task can rely on the :class:`.PropertySet` to communicate intermediate data among tasks.
+    """
+
+    @abstractmethod
+    def execute(
+        self,
+        passmanager_ir: PassManagerIR,
+        status: WorkflowStatus | None = None,
+        property_set: PropertySet | None = None,
+        callback: Callable = None,
+    ) -> PassManagerIR:
+        """Execute optimization task for input Qiskit IR.
+
+        Args:
+            passmanager_ir: Qiskit IR to optimize.
+            status: Mutable status object of the whole pass manager workflow.
+            property_set: A mutable data collection shared among all tasks.
+            callback: A callback function which is caller per execution of optimization task.
+
+        Returns:
+            Optimized Qiskit IR.
+        """
+        pass
+
+
+class GenericPass(Task, ABC):
+    """Base class of a single pass manager task.
+
+    A pass instance can read and write to the provided :class:`.PropertySet`,
+    and may modify the input pass manager IR.
+    """
+
+    def __init__(self):
+        self.requires: Iterable[Task] = []
+        self.property_set = PropertySet()
+
+    def name(self) -> str:
+        """Name of the pass."""
+        return self.__class__.__name__
+
+    def execute(
+        self,
+        passmanager_ir: PassManagerIR,
+        status: WorkflowStatus | None = None,
+        property_set: PropertySet | None = None,
+        callback: Callable = None,
+    ) -> PassManagerIR:
+        # Overriding this method is not safe.
+        # Pass subclass must keep current implementation.
+        # Especially, task execution may break when method signature is modified.
+
+        if property_set is not None:
+            self.property_set = property_set
+        if status is None:
+            status = WorkflowStatus()
+
+        for required in self.requires:
+            passmanager_ir = required.execute(
+                passmanager_ir=passmanager_ir,
+                status=status,
+                property_set=self.property_set,
+                callback=callback,
+            )
+
+        run_state = None
+        ret = None
+        start_time = time.time()
+        try:
+            if self not in status.completed_passes:
+                ret = self.run(passmanager_ir)
+                run_state = RunState.SUCCESS
+            else:
+                run_state = RunState.SKIP
+        except Exception as ex:
+            run_state = RunState.FAIL
+            raise ex
+        finally:
+            ret = ret or passmanager_ir
+            if run_state != RunState.SKIP:
+                running_time = time.time() - start_time
+                log_msg = f"Pass: {self.name()} - {running_time * 1000:.5f} (ms)"
+                logger.info(log_msg)
```

Comment:
We shouldn't pass an eagerly formatted string into the logger; the `logging.Logger` methods accept extra arguments so that the formatting only happens if the logger is enabled at that level.

You spell that something like
```python
logger.info("Pass: {} - {:.5f} (ms)", self.name(), running_time * 1000)
```
It's minor and in this case unlikely to have a notable slowdown (especially because we still make a function call and floating multiply!), it's just general logging practice - pylint would have warned about it if the f-string was inline in the call.

Reply:
I guess I just copied old code. Done in [f0a547a](https://github.com/Qiskit/qiskit/pull/10127/commits/f0a547a609378238ea32827418a810588cdbd522)